lib/datasets/downloader.rb

```diff
@@ -31,41 +33,59 @@ def download(output_path)
         headers["Range"] = "bytes=#{start}-"
       end
 
-      Net::HTTP.start(@url.hostname,
-                      @url.port,
-                      :use_ssl => (@url.scheme == "https")) do |http|
-        path = @url.path
-        path += "?#{@url.query}" if @url.query
-        request = Net::HTTP::Get.new(path, headers)
+      start_http(@url, headers) do |response|
+        case response
+        when Net::HTTPPartialContent
+          mode = "ab"
+        when Net::HTTPSuccess
+          start = nil
+          mode = "wb"
+        else
+          break
+        end
+
+        base_name = @url.path.split("/").last
+        size_current = 0
+        size_max = response.content_length
+        if start
+          size_current += start
+          size_max += start
+        end
+        progress_reporter = ProgressReporter.new(base_name, size_max)
+        partial_output_path.open(mode) do |output|
+          response.read_body do |chunk|
+            size_current += chunk.bytesize
+            progress_reporter.report(size_current)
+            output.write(chunk)
+          end
+        end
+      end
+      FileUtils.mv(partial_output_path, output_path)
+    end
+
+    private def start_http(url, headers, limit = 10, &block)
+      if limit == 0
+        raise TooManyRedirects, "too many redirections"
```

Comment:
How about showing URL?

```suggestion
        raise TooManyRedirects, "too many redirections: #{url}"
```

Reply:
This `url` is the last one to be tried with `limit = 0`. In other words, it is not the first `url` passed to `start_http` method.  It's OK?

Reply:
Hmm, can we show the first URL and the last URL?
Intermediate URLs are shown in `$stderr`, right?

Reply:
Yes, the intermediate URLs are shown in `$stderr`.
OK, I'll try to embed the first and the last URLs in the exception message.

Reply:
Implemented in https://github.com/red-data-tools/red-datasets/pull/99/commits/c36166e5f0111c5d459f7233cf99a6a665bab862
Could you check this change?

Reply:
We should move `RDatasets#filter` to `Dataset#filter` later. OK?

Reply:
Yes, I agree with you.